envoy/config/typed_config.h

```diff
@@ -32,6 +32,12 @@ class UntypedFactory {
    * @return configuration proto full name, or empty for untyped factories.
    */
   virtual std::string configType() { return ""; }
+
+  /**
+   * @return All full names of configuration protos used by the factories except the name
+   * returned by configType().
+   */
+  virtual std::set<std::string> otherTypes() { return {}; }
```

Comment:
I am a little bit confused about the `otherTypes`. What does it exactly do? It looks like the changes to `source/common/router/config_impl.cc` have realized the purpose of the patch.

Reply:
This method do the similar things as `configType()`. The returned type urls will be used when we build the factories map by type urls.

If no this method, we cannot find the correct factory by the route config proto.

Reply:
@mattklein123 It depends designs of extention factories. But usually, this shouldn't happened. The `otherTypes` is designed to be used by the factories that will consume multiple type proto configurations.

Reply:
If a factory takes multiple protos, we could just register it multiple times, right? I imagine this is useful for migrating between proto versions, but not seeing why this needs a special method.

Reply:
> If a factory takes multiple protos, we could just register it multiple times, right?

Yep.

> I imagine this is useful for migrating between proto versions.

This is designed for the factories which will consume multiple different proto configs such as http fitler factories. May be it's also useful for the mutiple versions of same factory? 🤔 

> but not seeing why this needs a special method.

This is to ensure this new feature will not break the exist extensions. But base on the latest investigate, it should be ok to update the exist `configType` to return a set of type url. Check https://github.com/envoyproxy/envoy/pull/21030#issuecomment-1112813839